File: app/src/main/java/me/blog/korn123/easydiary/adapters/DiaryMainItemAdapter.java
package me.blog.korn123.easydiary.adapters;

import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.GradientDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.apache.commons.lang3.StringUtils;

import java.util.List;

import me.blog.korn123.easydiary.extensions.ContextKt;
import me.blog.korn123.easydiary.extensions.ContextKt.*;
import io.github.hanjoongcho.commons.helpers.BaseConfig;
import me.blog.korn123.commons.constants.Constants;
import me.blog.korn123.commons.utils.CommonUtils;
import me.blog.korn123.commons.utils.DateUtils;
import me.blog.korn123.commons.utils.EasyDiaryUtils;
import me.blog.korn123.commons.utils.FontUtils;
import me.blog.korn123.easydiary.R;
import me.blog.korn123.easydiary.models.DiaryDto;

/**
 * Created by CHO HANJOONG on 2017-03-16.
 */

public class DiaryMainItemAdapter extends ArrayAdapter<DiaryDto> {
    private final Context context;
    private final List<DiaryDto> list;
    private final int layoutResourceId;
    private String query;

    public void setCurrentQuery(String query) {
        this.query = query;
    }

    public String getCurrentQuery() {
        return this.query;
    }

    public DiaryMainItemAdapter(Context context, int layoutResourceId, List<DiaryDto> list) {
        super(context, layoutResourceId, list);
        this.context = context;
        this.list = list;
        this.layoutResourceId = layoutResourceId;
    }

    public View getView(int position, View convertView, ViewGroup parent) {
        View row = convertView;
        ViewHolder holder = null;
        if (row == null) {
            LayoutInflater inflater = ((Activity)this.context).getLayoutInflater();
            row = inflater.inflate(this.layoutResourceId, parent, false);
            holder = new ViewHolder();
            holder.textView1 = row.findViewById(R.id.text1);
            holder.textView2 = row.findViewById(R.id.text2);
            holder.textView3 = row.findViewById(R.id.text3);
            holder.imageView = row.findViewById(R.id.weather);
            holder.item_holder = row.findViewById(R.id.item_holder);
            row.setTag(holder);
        } else {
            holder = (ViewHolder)row.getTag();
        }

        setFontsTypeface(holder);
        setFontsSize(holder);

        DiaryDto diaryDto = (DiaryDto)this.list.get(position);
        if (StringUtils.isEmpty(diaryDto.getTitle())) {
            holder.textView1.setVisibility(View.GONE);
        } else {
            holder.textView1.setVisibility(View.VISIBLE);
        }
        holder.textView1.setText(diaryDto.getTitle());
        holder.textView2.setText(diaryDto.getContents());

        // highlight current query
        if (StringUtils.isNotEmpty(query)) {
            if (CommonUtils.loadBooleanPreference(getContext(), Constants.DIARY_SEARCH_QUERY_CASE_SENSITIVE)) {
                EasyDiaryUtils.highlightString(holder.textView1, query);
                EasyDiaryUtils.highlightString(holder.textView2, query);
            } else {
                EasyDiaryUtils.highlightStringIgnoreCase(holder.textView1, query);
                EasyDiaryUtils.highlightStringIgnoreCase(holder.textView2, query);
            }

        }
        holder.textView3.setText(DateUtils.getFullPatternDateWithTime(diaryDto.getCurrentTimeMillis()));
        EasyDiaryUtils.initWeatherView(holder.imageView, diaryDto.getWeather());

        GradientDrawable drawable = (GradientDrawable) holder.item_holder.getBackground();
        drawable.setColor(new BaseConfig(context).getCustomBackgroundColor());
        ContextKt.updateTextColors(context, holder.item_holder, 0, 0);
        return row;
    }

    private void setFontsTypeface(ViewHolder holder) {
        FontUtils.setFontsTypeface(context, context.getAssets(), null, holder.textView1, holder.textView2, holder.textView3);
    }

    private void setFontsSize(ViewHolder holder) {
        float commonSize = CommonUtils.loadFloatPreference(context, Constants.SETTING_FONT_SIZE, holder.textView1.getTextSize());
        FontUtils.setFontsSize(commonSize, -1, holder.textView1, holder.textView2, holder.textView3);
    }

    private static class ViewHolder {
        TextView textView1;
        TextView textView2;
        TextView textView3;
        ImageView imageView;
        LinearLayout item_holder;
    }
}

File: app/src/main/java/me/blog/korn123/easydiary/adapters/DiaryMainItemAdapter.kt
package me.blog.korn123.easydiary.adapters

import android.animation.ArgbEvaluator
import android.app.Activity
import android.graphics.drawable.Drawable
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView
import io.noties.markwon.Markwon
import io.noties.markwon.movement.MovementMethodPlugin
import me.blog.korn123.commons.utils.DateUtils
import me.blog.korn123.commons.utils.EasyDiaryUtils
import me.blog.korn123.commons.utils.EasyDiaryUtils.createBackgroundGradientDrawable
import me.blog.korn123.commons.utils.EasyDiaryUtils.createThumbnailGlideOptions
import me.blog.korn123.commons.utils.FlavorUtils
import me.blog.korn123.commons.utils.FontUtils
import me.blog.korn123.easydiary.R
import me.blog.korn123.easydiary.activities.DiaryMainActivity
import me.blog.korn123.easydiary.databinding.ItemDiaryMainBinding
import me.blog.korn123.easydiary.enums.Calculation
import me.blog.korn123.easydiary.enums.DiaryMode
import me.blog.korn123.easydiary.extensions.applyMarkDownPolicy
import me.blog.korn123.easydiary.extensions.changeDrawableIconColor
import me.blog.korn123.easydiary.extensions.config
import me.blog.korn123.easydiary.extensions.dpToPixel
import me.blog.korn123.easydiary.extensions.initTextSize
import me.blog.korn123.easydiary.extensions.updateAppViews
import me.blog.korn123.easydiary.extensions.updateCardViewPolicy
import me.blog.korn123.easydiary.extensions.updateTextColors
import me.blog.korn123.easydiary.helper.EasyDiaryDbHelper
import me.blog.korn123.easydiary.helper.PHOTO_CORNER_RADIUS_SCALE_FACTOR_NORMAL
import me.blog.korn123.easydiary.helper.THUMBNAIL_BACKGROUND_ALPHA
import me.blog.korn123.easydiary.models.Diary
import org.apache.commons.lang3.StringUtils

class DiaryMainItemAdapter(
    val activity: Activity,
    private val diaryItems: List<Diary>,
    val itemClickCallback: (diary: Diary) -> Unit,
    val itemLongClickCallback: () -> Unit
) : RecyclerView.Adapter<DiaryMainItemAdapter.ViewHolder>(),
    FastScrollRecyclerView.SectionedAdapter {
    var currentQuery: String? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            activity,
            ItemDiaryMainBinding.inflate(activity.layoutInflater, parent, false),
            this
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindTo(diaryItems[position])
    }

    override fun getItemCount(): Int = diaryItems.size

    override fun getSectionName(position: Int): String {
        fun ellipsis(str: String?): String {
            val maxLength = 15
            val ellipsisString = str?.let {
                when (str.length > maxLength) {
                    true -> str.take(maxLength).plus("…")
                    false -> str
                }
            } ?: ""
            return ellipsisString
        }

        val label = when (diaryItems[position].title?.isNotEmpty() ?: false) {
            true -> String.format("%s", ellipsis(diaryItems[position].title))
            false -> String.format("%s", ellipsis(diaryItems[position].contents))
        }
        return label
    }

    fun getSelectedItems(): List<Diary> {
        val selectedItems = arrayListOf<Diary>()
        diaryItems.map {
            if (it.isSelected) selectedItems.add(it)
        }
        return selectedItems
    }

    inner class ViewHolder(
        val activity: Activity,
        private val itemDiaryMainBinding: ItemDiaryMainBinding,
        val adapter: DiaryMainItemAdapter
    ) : RecyclerView.ViewHolder(itemDiaryMainBinding.root) {
        fun bindTo(diary: Diary) {
            itemDiaryMainBinding.run {
                activity.run {
                    root.run {
                        setOnClickListener { itemClickCallback.invoke(diary) }
                        setOnLongClickListener {
                            itemLongClickCallback()
                            true
                        }
                        updateTextColors(this, 0, 0)
                        updateAppViews(this)
                        initTextSize(this)
                        updateCardViewPolicy(this)
                        FontUtils.setFontsTypeface(context, null, this)
                    }

                    if (config.enableLocationInfo) {
                        diary.location?.let {
                            changeDrawableIconColor(config.primaryColor, R.drawable.ic_map_marker_2)
//                        locationLabel.setTextColor(config.textColor)
//                        locationContainer.background = getLabelBackground()

                            locationLabel.text = it.address
                            locationContainer.visibility = View.VISIBLE
                        } ?: run {
                            locationContainer.visibility = View.GONE
                        }
                    } else {
                        locationContainer.visibility = View.GONE
                    }

                    if (config.enableCountCharacters) {
                        contentsLength.run {
//                        setTextColor(config.textColor)
//                        background = getLabelBackground()

                            text = context.getString(
                                R.string.diary_contents_length,
                                diary.contents?.length ?: 0
                            )
                        }
                        contentsLengthContainer.visibility = View.VISIBLE
                    } else {
                        contentsLengthContainer.visibility = View.GONE
                    }
                }

                selection.setOnCheckedChangeListener { _, isChecked ->
                    EasyDiaryDbHelper.beginTransaction()
                    diary.isSelected = isChecked
                    EasyDiaryDbHelper.commitTransaction()
                    // EasyDiaryDbHelper.updateDiaryBy(diary)
                }

                when ((activity as DiaryMainActivity).mDiaryMode) {
                    DiaryMode.READ -> selection.visibility = View.GONE
                    DiaryMode.DELETE -> {
                        selection.visibility = View.VISIBLE
                        selection.isChecked = diary.isSelected
                    }
                }

                if (StringUtils.isEmpty(diary.title)) {
                    textTitle.visibility = View.GONE
                } else {
                    textTitle.visibility = View.VISIBLE
                }
                textTitle.text = diary.title
                activity.applyMarkDownPolicy(textContents, diary.contents!!)

                // highlight current query
                if (StringUtils.isNotEmpty(currentQuery)) {
                    val color = ArgbEvaluator().evaluate(0.75F, 0x000000, 0xffffff) as Int
                    if (activity.config.diarySearchQueryCaseSensitive) {
                        EasyDiaryUtils.highlightString(textTitle, currentQuery)
                        EasyDiaryUtils.highlightString(textContents, currentQuery)
                    } else {
                        EasyDiaryUtils.highlightStringIgnoreCase(textTitle, currentQuery)
                        EasyDiaryUtils.highlightStringIgnoreCase(textContents, currentQuery)
                    }

                }
                EasyDiaryUtils.boldString(activity, textTitle)

                textDateTime.text = when (diary.isAllDay) {
                    true -> DateUtils.getDateStringFromTimeMillis(diary.currentTimeMillis)
                    false -> DateUtils.getDateTimeStringFromTimeMillis(diary.currentTimeMillis)
                }
                if (activity.config.enableDebugMode) textDateTime.text =
                    "[${diary.originSequence}] ${textDateTime.text}"
                FlavorUtils.initWeatherView(activity, imageSymbol, diary.weather)

                when ((diary.photoUris?.size ?: 0) > 0) {
                    true -> {
                        photoViews.visibility = View.VISIBLE
                    }

                    false -> photoViews.visibility = View.GONE
                }

                photoViews.removeAllViews()
                if ((diary.photoUris?.size ?: 0) > 0) {
                    diary.photoUrisWithEncryptionPolicy()?.map {
                        val imageXY = activity.dpToPixel(32F)
                        val imageView = ImageView(activity)
                        val layoutParams = LinearLayout.LayoutParams(imageXY, imageXY)
                        layoutParams.setMargins(
                            0,
                            activity.dpToPixel(1F),
                            activity.dpToPixel(3F),
                            0
                        )
                        imageView.layoutParams = layoutParams
                        imageView.background = createBackgroundGradientDrawable(
                            activity.config.primaryColor,
                            THUMBNAIL_BACKGROUND_ALPHA,
                            imageXY * PHOTO_CORNER_RADIUS_SCALE_FACTOR_NORMAL
                        )
                        imageView.scaleType = ImageView.ScaleType.CENTER
                        activity.dpToPixel(1.5F, Calculation.FLOOR).apply {
                            imageView.setPadding(this, this, this, this)
                        }
                        val listener = object : RequestListener<Drawable> {
                            override fun onLoadFailed(
                                e: GlideException?,
                                model: Any?,
                                target: Target<Drawable>?,
                                isFirstResource: Boolean
                            ): Boolean {
                                return false
                            }

                            override fun onResourceReady(
                                resource: Drawable?,
                                model: Any?,
                                target: Target<Drawable>?,
                                dataSource: DataSource?,
                                isFirstResource: Boolean
                            ): Boolean {
                                return false
                            }
                        }
                        Glide.with(activity)
                            .load(EasyDiaryUtils.getApplicationDataDirectory(activity) + it.getFilePath())
                            .listener(listener)
                            .apply(
                                createThumbnailGlideOptions(
                                    imageXY * PHOTO_CORNER_RADIUS_SCALE_FACTOR_NORMAL,
                                    it.isEncrypt()
                                )
                            )
                            .into(imageView)
//                    if (photoViews.childCount >= maxPhotos) return@map
                        photoViews.addView(imageView)
                    }
                }

                textContents.maxLines = when (activity.config.enableContentsSummary) {
                    true -> activity.config.summaryMaxLines
                    false -> Integer.MAX_VALUE
                }
            }
        }
    }
}